ChildMindInstitute/mindlogger-admin
# feat: Unassign Activities (M2-7277, M2-7547)

### 📝 Description

🔗 [Jira Ticket M2-7277](https://mindlogger.atlassian.net/browse/M2-7277): Unassign Activity or Flow
🔗 [Jira Ticket M2-7547](https://mindlogger.atlassian.net/browse/M2-7547): Unassign Tooltip/Copy for Auto-Assigned Activities

This PR adds the ability to unassign respondents and target subjects from assigned activities/flows on the Participant Details Page (PDP):

- Adds the "Unassign Activity" menu item to assigned activity/flow cards on the PDP
- Adds the "Unassigned" section to the bottom of the PDP to list unassigned activities/flows
- When "Unassign Activity" is clicked, displays the [Unassign dialogue](https://www.figma.com/design/6BNx63peNVzrHAF1FoZ78x/Multi-Informant-2024?node-id=11807-162282&node-type=frame&t=z8cp3iGbC7PeNhEy-11) (if the activity/flow has only one assignment) or the [Unassign panel](https://www.figma.com/design/6BNx63peNVzrHAF1FoZ78x/Multi-Informant-2024?node-id=11807-265109&node-type=frame&t=z8cp3iGbC7PeNhEy-11) (if the activity/flow has multiple assignments).
- Confirming the unassign action calls the new DELETE `/assignments/applet/{applet_id}` endpoint to remove the assignment(s) and re-queries the PDP

> [!NOTE]
> The [Unassign dialogue](https://www.figma.com/design/6BNx63peNVzrHAF1FoZ78x/Multi-Informant-2024?node-id=12187-395534&node-type=frame&t=z8cp3iGbC7PeNhEy-11) & [Unassign panel](https://www.figma.com/design/6BNx63peNVzrHAF1FoZ78x/Multi-Informant-2024?node-id=11807-276493&node-type=frame&t=z8cp3iGbC7PeNhEy-11) are based on the new Unassign flow from the [new (revised) PDP designs](https://www.figma.com/design/6BNx63peNVzrHAF1FoZ78x/Multi-Informant-2024?node-id=11796-234679&node-type=frame&t=z8cp3iGbC7PeNhEy-11). However, the changes made to the PDP itself in this PR to identify unassigned activities (adding the "Unassigned" section) are based on the [original designs](https://www.figma.com/design/6BNx63peNVzrHAF1FoZ78x/Multi-Informant-2024?node-id=9135-121172&t=z8cp3iGbC7PeNhEy-4). Actual revisions to the PDP to use the newest designs were out of scope for this ticket, and will be happening subsequently this sprint.

> [!TIP]
> Related changes have been grouped by commit, with detailed descriptions of why certain changes were made in commit messages. It will probably be easier to review [commit-by-commit](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1924/commits).

### 📸 Screenshots


https://github.com/user-attachments/assets/a99d40f2-9056-49bb-91f7-1d18c25633cc


### 🪤 Peer Testing

> [!NOTE] 
> This feature depends on the `enableActivityAssign` feature flag. This should already be enabled in dev/local environments. To control its value manually, add `features.enableActivityAssign = [true|false]` prior to the `return` statement [here in `useFeatureFlags`](https://github.com/ChildMindInstitute/mindlogger-admin/blob/e6ff173fc6aaf31da3e81d92ab4f59caa8eef6da/src/shared/hooks/useFeatureFlags.ts#L25-L26).

#### Test with feature flag enabled ✅

1. Add `features.enableActivityAssign = true` to [`useFeatureFlags`](https://github.com/ChildMindInstitute/mindlogger-admin/blob/e6ff173fc6aaf31da3e81d92ab4f59caa8eef6da/src/shared/hooks/useFeatureFlags.ts#L25-L26).
2. Please refer to the AC of [M2-7277](https://mindlogger.atlassian.net/browse/M2-7277) and [M2-7547](https://mindlogger.atlassian.net/browse/M2-7547) to ensure behaviour is in alignment.

#### Test with feature flag disabled 🚫

1. Add `features.enableActivityAssign = false` to [`useFeatureFlags`](https://github.com/ChildMindInstitute/mindlogger-admin/blob/e6ff173fc6aaf31da3e81d92ab4f59caa8eef6da/src/shared/hooks/useFeatureFlags.ts#L25-L26).
2. Load the PDP and ensure there are is regression in behaviour.

### ✏️ Notes

- Refactored the internal banners hook previously used by `ActivityAssignDrawer` to a shared hook called `useModalBanners`, which gives any component (drawer or otherwise) the convenience to render its custom set of banners internally.
- Hid a useless Search field on the PDP that was never actually implemented and should have been commented out with a TODO link long ago.
- Removed some console warnings by avoiding use of `:first-child` in CSS.

[M2-7277]: https://mindlogger.atlassian.net/browse/M2-7277?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
